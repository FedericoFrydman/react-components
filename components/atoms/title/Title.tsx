import { forwardRef, Ref } from 'react';
import cn from 'classnames';

import { Heading } from '../heading/Heading';
import { Text, TextProps } from '../text/Text';
import '../../assets/styles/utilities.scss';

export type TitleSize = 'small' | 'medium' | 'large';

export interface TitleProps extends Omit<TextProps, 'textSize'> {
  /**
   * Size of the title. Defaults to `medium`.
   */
  textSize?: TitleSize;
  /**
   * Custom classes for the component.
   */
  classNames?: cn.Argument;
  /**
   * The ref to the HTML DOM element.
   */
  ref?: Ref<HTMLParagraphElement>;
}

const TitleComponent = ({ text, textSize, classNames = [], ...props }: TitleProps, ref: TitleProps['ref']) => {
  switch (textSize) {
    case 'large':
      return <Heading ref={ref} text={text} as="T20" classNames={classNames} />;
    case 'small':
      return <Text {...props} ref={ref} textSize="small" text={text} classNames={classNames} />;
    case 'medium':
    default:
      return <Text {...props} ref={ref} textSize="medium" text={text} classNames={classNames} />;
  }
};

export const Title = forwardRef<HTMLParagraphElement, TitleProps>(TitleComponent);
