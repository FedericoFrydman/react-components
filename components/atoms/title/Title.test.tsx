import { createRef } from 'react';
import { render, screen } from '@testing-library/react';
import { shallow, ShallowWrapper } from 'enzyme';
import cn from 'classnames';

import { Title, TitleProps } from './Title';

describe('Title Component', () => {
  let testId: string;
  let wrapper: ShallowWrapper;

  beforeEach(() => {
    wrapper = shallow(<Title text="some text" />);
    testId = 'test-text';
  });

  it('should render the component', () => {
    expect(wrapper).toMatchSnapshot();
  });

  describe('props', () => {
    describe('classNames', () => {
      describe('when provided classNames', () => {
        const classNames = ['my-class-1', 'my-class-2'];

        it('gives the text the provided classNames', () => {
          const text = renderTitle({ classNames });

          expect(text).toHaveClass(cn(classNames), { exact: false });
        });
      });
    });

    describe('text', () => {
      const text = 'testing text';

      it('is displayed when present', () => {
        renderTitle({ text });

        expect(screen.getByText(text)).toBeVisible();
      });
    });
  });

  describe('ref', () => {
    const ref = createRef<HTMLParagraphElement>();

    it('provides a ref to the text', () => {
      expect(renderTitle({ ref })).toEqual(ref.current);
    });
  });

  const renderTitle = (props: TitleProps = {}): HTMLElement => {
    render(<Title {...props} data-testid={testId} />);

    return screen.getByTestId(testId);
  };
});
