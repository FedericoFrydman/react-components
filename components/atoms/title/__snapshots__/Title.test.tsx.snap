// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Title Component should render the component 1`] = `
<ForwardRef(TextComponent)
  classNames={Array []}
  text="some text"
  textSize="medium"
/>
`;
